### PR TITLE
Cleaned up metadata code

### PR DESCRIPTION
**Pull request by harlanhaskins**
- Removed C-style casts in favor of `reinterpret_cast`
- Added documentation comments to the public metadata functions.
- Added `const` qualifiers to all types that can be `const`.
- Re-declared structs as C++-style structs if they don't need to be visible from Trill.
- Explicitly specified types inside those structs instead of `void *`.

### CODE REVIEW

File: Sources/Runtime/generics.cpp [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 using namespace trill;
 
-void *trill_createGenericBox(void *typeMetadata, void *witnessTable) {
+void *trill_createGenericBox(const void *typeMetadata, const void **witnessTable) {
     trill_assert(typeMetadata != nullptr);
     trill_assert(witnessTable != nullptr);
     auto metadata = (TypeMetadata *)typeMetadata;
```
Comment (segiddins) [inline Sources/Runtime/generics.cpp]: `reinterpret_cast`

File: Sources/Runtime/generics.cpp [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 using namespace trill;
 
-void *trill_createGenericBox(void *typeMetadata, void *witnessTable) {
+void *trill_createGenericBox(const void *typeMetadata, const void **witnessTable) {
```
Comment (segiddins) [inline Sources/Runtime/generics.cpp:7]: why the change to `void **`

File: Sources/Runtime/metadata.cpp [hunk 3]
```diff
@@ -129,35 +125,39 @@ TRILL_ANY trill_extractAnyField(TRILL_ANY any_, uint64_t fieldNum) {
...
   }
   auto fieldPtr = trill_getAnyFieldValuePtr({any}, fieldNum);
   auto newPtr = trill_getAnyValuePtr({newAny});
   memcpy(fieldPtr, newPtr, newType->sizeInBits);
 }
 
-void *trill_getAnyValuePtr(TRILL_ANY anyValue) {
-  return (void *)((intptr_t)anyValue._any + sizeof(AnyBox));
+void *_Nonnull trill_getAnyValuePtr(TRILL_ANY anyValue) {
+  return reinterpret_cast<void *>((intptr_t)anyValue._any + sizeof(AnyBox));
```
Comment (segiddins) [inline Sources/Runtime/metadata.cpp]: cast

**Review by segiddins — COMMENTED** [review]
this looks amazing!

File: Sources/Runtime/generics.cpp (hunk 2 above)
Comment (harlanhaskins) [reply Sources/Runtime/generics.cpp:7]: I realized it is actually a `void **`. It's a pointer to function pointers. It doesn't need a count, though, because accesses into this table will be done at compile time.

File: Sources/Runtime/generics.cpp (hunk 1 above)
Comment (harlanhaskins) [reply Sources/Runtime/generics.cpp]: Good catch!

File: Sources/Runtime/metadata.cpp (hunk 3 above)
Comment (harlanhaskins) [reply Sources/Runtime/metadata.cpp]: Good catch!

**Comment by segiddins** [comment]
```diff
diff --git a/trill.xcodeproj/project.pbxproj b/trill.xcodeproj/project.pbxproj
index 1753d4d..77f6728 100644
--- a/trill.xcodeproj/project.pbxproj
+++ b/trill.xcodeproj/project.pbxproj
@@ -1565,6 +1565,10 @@
 				MACOSX_DEPLOYMENT_TARGET = 10.12;
 				MTL_ENABLE_DEBUG_INFO = YES;
 				ONLY_ACTIVE_ARCH = YES;
+				OTHER_CPLUSPLUSFLAGS = (
+					"$(OTHER_CFLAGS)",
+					"-Wold-style-cast",
+				);
 				SDKROOT = macosx;
 				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
 			};
@@ -1607,6 +1611,10 @@
 				GCC_WARN_UNUSED_VARIABLE = YES;
 				MACOSX_DEPLOYMENT_TARGET = 10.12;
 				MTL_ENABLE_DEBUG_INFO = NO;
+				OTHER_CPLUSPLUSFLAGS = (
+					"$(OTHER_CFLAGS)",
+					"-Wold-style-cast",
+				);
 				SDKROOT = macosx;
 			};
 			name = Release;
```

**Review by segiddins — APPROVED** [review]
